src/coreclr/jit/codegenxarch.cpp

```diff
@@ -4717,7 +4717,17 @@ void CodeGen::genCodeForIndir(GenTreeIndir* tree)
     else
     {
         genConsumeAddress(addr);
-        emit->emitInsLoadInd(ins_Load(targetType), emitTypeSize(tree), tree->GetRegNum(), tree);
+        instruction loadIns = ins_Load(targetType);
+        if (tree->DontExtend())
+        {
+            assert(varTypeIsSmall(tree));
+            // The user of this IND does not need
+            // the upper bits to be set, so we don't need to use longer
+            // INS_movzx/INS_movsx and can use INS_mov instead.
+            // It usually happens when the real type is a small struct.
+            loadIns = INS_mov;
+        }
+        emit->emitInsLoadInd(loadIns, emitTypeSize(tree), tree->GetRegNum(), tree);
```

Comment:
This logic allows us to generate `STORE_LCL_VAR struct<1> V01(IND struct(addr)` as
```
mov al, [addr]
mov  [V11], al
```
instead of 
```
movsx rax, [addr]
mov  [V11], al
```

The solution with additional flag does not look very nice but it was better than alternatives that I have tried:
1. support contained IND under `STORE_LCL_VAR`: it required too many changes in LSRA/codegen/emitter and it was hard to make it work for all cases and it created many new checks for  this scenario;
2. type the `IND` as `int` so we do `mov rax, [addr]` instead of `mov al, [addr]`. This could be a solution without any additional diffs but I am not sure if this transformation is correct. 
We currently use it for nullchecks already since https://github.com/dotnet/runtime/pull/37991/files so `NULLCHECK byte addr` is currently  transformed into `NULLCHECK int addr` but if the `addr` is the last byte on a last readable page won't it cause issues? Maybe we should reconsider this transformation (in https://github.com/dotnet/runtime/blob/78593b9e095f974305b2033b465455e458e30267/src/coreclr/jit/lower.cpp#L6736)
could somebody please confirm if it can't cause issues?

![image](https://user-images.githubusercontent.com/22642771/133939442-3e374064-20c5-48b6-8f68-31d1de1c0a91.png)